### PR TITLE
Fix wrong page title and missing head tags on login

### PR DESCRIPTION
**Pull request by Tug**
Fixes https://github.com/Automattic/wp-calypso/issues/18919

https://github.com/Automattic/wp-calypso/pull/17134 introduced a new bug because of `setRouteMiddleware` called for isomorphic sections. It would dispatch a `ROUTE_SET` action which would reset the page title and meta as defined in the reducer https://github.com/Automattic/wp-calypso/blob/6ec964c85d93bdc7f63400e573a7df6babd64e51/client/state/document-head/reducer.js#L26

While investigating a fix, I noticed another [problem with the `cacheQueryKeys`](https://github.com/Automattic/wp-calypso/pull/19036#issuecomment-341728386) feature. 

In order to fix both issues, the solution adopted in this PR is simple, refactor the caching part of the state so we:
- always start with an empty state when creating the initial context
- cache the state only after a page is rendered with SSR (we don't need to cache the state in other cases)
- retrieve the state and the rendered page at the same time (and overwrite the existing store).

This approach has a pros:
- It's cleaner to always start with an empty state and it makes reasoning with SSR simpler
- It fixes our issue with `ROUTE_SET` and any potential action dispatched on a pre-rendered route that would modify the state.
- It fixes the `context.cacheQueryKeys` feature as it will always be defined when `context.layout` is set.

And cons:
- middlewares that bind the store (`context.store.subscribe`) will cease to work after we replace `context.store`. It's not a problem currently as rendering is the last middleware in the stack and there shouldn't be any code that listen for changes in the state on the server.

### Testing Instructions

You should execute the following `curl` commands **at least two times** in order to check the behavior when the cache is empty (first page load), and filled (subsequent page loads):

1. Run `git checkout fix/ssr-head` and start your server
2. Assert that the English `Login` page has the right title:
```
curl -s http://calypso.localhost:3000/log-in | grep -o '<title>[^<]*</title>'
<title>Log In — WordPress.com</title>
```
3. Assert that non-English `Login` pages have a default title:
```
curl -s http://calypso.localhost:3000/log-in/fr | grep -o '<title>[^<]*</title>'
<title>WordPress.com</title>
```
4. Assert that the English `Login` page has a `canonical` tag:
```
curl -s http://calypso.localhost:3000/log-in | grep -o '<link rel="canonical"[^>]*>'
<link rel="canonical" href="https://wordpress.com/login">
```
5. Assert that non-English `Login` pages don't have this tag:
```
curl -s http://calypso.localhost:3000/log-in/he | grep -o '<link rel="canonical"[^>]*>'
```
6. Assert that the WooCommerce `Login` page has the right title:
```
curl -s "http://calypso.localhost:3000/log-in?client_id=50916&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fresponse_type%3Dcode%26client_id%3D50916%26state%3Dce55c88fc2fc12dadf7c12874e4178ab%26redirect_uri%3Dhttps%253A%252F%252Fwoocommerce.com%252Fwc-api%252Fwpcom-signin%253Fnext%253Dmy-dashboard%26blog_id%3D0%26wpcom_connect%3D1" | grep -o '<title>[^<]*</title>'
<title>Log In — WordPress.com</title>
```
And that it contains the WooCommerce logo:
```
curl -s "http://calypso.localhost:3000/log-in?client_id=50916&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fresponse_type%3Dcode%26client_id%3D50916%26state%3Dce55c88fc2fc12dadf7c12874e4178ab%26redirect_uri%3Dhttps%253A%252F%252Fwoocommerce.com%252Fwc-api%252Fwpcom-signin%253Fnext%253Dmy-dashboard%26blog_id%3D0%26wpcom_connect%3D1" | grep -o '<img[^>]*logo-woocommerce@2x.png'
<img src="https://woocommerce.com/wp-content/themes/woo/images/logo-woocommerce@2x.png
```

### Reviews
- [x] Code
- [x] Product

### CODE REVIEW

**Comment by Tug** [comment]
I still need to double check that persisting `state.ui` won't cause problems. But otherwise I think it's ready for a review!

File: server/isomorphic-routing/index.js [hunk 1]
```diff
@@ -45,10 +47,14 @@ export function serverRouter( expressApp, setUpRoute, section ) {
 }
 
 function setRouteMiddleware( context, next ) {
-	context.store.dispatch( setRouteAction(
-		context.pathname,
-		context.query
-	) );
+	const state = context.store.getState();
+
+	if ( context.pathname !== getCurrentRoute( state ) || ! isEqual( context.query, getCurrentQueryArguments( state ) ) ) {
```
Comment (yurynix) [inline server/isomorphic-routing/index.js]: super minor: I would prefer an explicit selector that does that:

```js
if ( ! isCurrentRoute( context.pathname, context.query ) ) { ... }
```

File: server/isomorphic-routing/index.js [hunk 2]
```diff
@@ -96,10 +102,18 @@ function compose( ...functions ) {
 }
 
 export function getCacheKey( context ) {
+	// express.js Request object has a path property for the pathname (path without the querystring)
```
Comment (yurynix) [inline server/isomorphic-routing/index.js]: Maybe it's just me, but I'm super confused by this function.

I'm not sure what we receive here as an argument,
I mean, we call it here `context`, but:

[here we pass](https://github.com/Automattic/wp-calypso/blob/5c1ba6433c69a6a35aec23a588f42344031e511c/server/pages/index.js#L146) express request
[here we pass](https://github.com/Automattic/wp-calypso/blob/2e0452a50b013a60642e3f8abf799f4587466f61/server/render/index.js#L85) context

IMO the argument named `context` here needs a different naming, maybe even change signature.

Also, if we can path either a request or either request how does this work when we pass request:
```
if ( isEmpty( context.query ) || isEmpty( context.cacheQueryKeys ) ) { ... }
```

Is that supposed to detect whether that's a context object or a request object?

TL;DR:
* We should make super clear what this function can receive from it's signature / jsdoc
* If we're trying to detect what is actually received ( context or request ) lets say why we're doing it ( we don't cache that? we cache it? we ignore it? )

**Comment by yurynix** [comment]
Thank you for this Tug, Code looks great! Also a nice refactor.
I had a minor comment about the `getCacheKey` function I think should be addressed before merging. I can do it myself if you'd like, just ping me.

**Comment by Tug** [comment]
So, I had to refactor the way we cache state because it was not actually working.
Currently we're using `getCacheKey()` to compute a key given a route (with or without query args). The problem is for routes that use `cacheQueryKeys`. The caching part works well, we can cache `/log-in?cliend_id=930` without any issue. But then when we try to access this page again, because `context.cacheQueryKeys` is only set from a client middleware (during the rendering part), it will always be empty. So the cache key for this route when we try to get the initial redux will always be `/log-in` on subsequent calls.
To fix that, I have decided to group together component and state rendering and always start with an empty state. Only when we retrieve the cached layout will we also overwrite the redux store data. It should work in most cases as all the existing middlewares that access the state before rendering use it to prepare for rendering (I haven't seen any redirection or special cases so far).

It's a change so we should do some testing on SSR routes, especially those that use query parameters.

**Comment by yurynix** [comment]
Not really sure how the last check was supposed to work since it redirects to WordPress.com...
I used instead:
```
curl --resolve calypso.localhost:3000:127.0.0.1 http://calypso.localhost:3000/log-in?client_id=50916&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fresponse_type%3Dcode%26client_id%3D50916%26state%3D1c5346619ebad1f69d80f406709d6e42%26redirect_uri%3Dhttps%253A%252F%252Fwoocommerce.com%252Fwc-api%252Fwpcom-signin%253Fnext%253Dmy-dashboard%26blog_id%3D0%26wpcom_connect%3D1 | grep "logo-woocommerce@2x.png"
```

File: client/state/selectors/get-current-query-arguments.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Eternal dependencies
```
Comment (yurynix) [inline client/state/selectors/get-current-query-arguments.js]: typo

File: client/state/selectors/get-current-route.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Eternal dependencies
```
Comment (yurynix) [inline client/state/selectors/get-current-route.js]: typo

File: client/state/selectors/get-initial-query-arguments.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Eternal dependencies
```
Comment (yurynix) [inline client/state/selectors/get-initial-query-arguments.js]: typo

File: client/state/selectors/get-current-route.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+
+/**
+ * Gets the last route set by a ROUTE_SET action
+ * @param {Object} state - global redux state
+ * @return {string} current state value
```
Comment (yurynix) [inline client/state/selectors/get-current-route.js]: Maybe
```
@return {string} current route value
```
?

File: client/state/selectors/get-initial-query-arguments.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+
+/**
+ * Gets the initial query parameters set by a ROUTE_SET action
+ * @param {Object} state - global redux state
+ * @return {Object} current state value
```
Comment (yurynix) [inline client/state/selectors/get-initial-query-arguments.js]: Maybe
```
@return {object} initial query arguments
```

File: server/render/index.js [hunk 8]
```diff
@@ -40,25 +41,64 @@ function bumpStat( group, name ) {
...
+				// Send state to client
+				context.initialReduxState = pick( context.store.getState(), reduxSubtrees );
+				const serverState = reducer( context.initialReduxState, { type: SERIALIZE } );
+				stateCache.set( key, serverState );
+			}
+		} else {
+			const serializedServerState = stateCache.get( key );
+			if ( serializedServerState ) {
+				context.initialReduxState = reducer( serializedServerState, { type: DESERIALIZE } );
+				context.store = createReduxStore( context.initialReduxState );
```
Comment (yurynix) [inline server/render/index.js]: Why this is required?
I don't see anything using that further down the line...

**Comment by yurynix** [comment]
Thank you for working on this @Tug !
It works as described.
I had only one minor question and I think that's almost ready to go.

File: server/render/index.js (hunk 8 above)
Comment (Tug) [reply server/render/index.js]: All those access the store though:
https://github.com/Automattic/wp-calypso/pull/19036/files#diff-d92d930a21711a960b300aeb231d9056R135

Comment (yurynix) [reply server/render/index.js]: Aha, I see, thank you!
It's not that optimal to initialize the store just for that, but lets merge and optimize later...

**Comment by Tug** [comment]
Looks good, thanks for your contributions!

**Comment by stephanethomas** [comment]
I've updated the testing instructions to check the original issue described https://github.com/Automattic/wp-calypso/issues/18919, i.e. the page has the right title on the first load but a wrong one for any subsequent page loads. I did also update the `curl` commands to only display the content of the tags searched (as otherwise they would return a wall of HTML code) and to actually test the local `Login` page for WooCommerce.

**Comment by ockham** [comment]
Are you folks sure we have to essentially rewrite everything related to isomorphic routing? I had only a quick glance, but the amount of changes to the `server/` directory makes me a bit uneasy. I'd love to review this PR, but the amount of rewriting going on here doesn't make it exactly easy, and the PR desc isn't that enlightening for me.

> https://github.com/Automattic/wp-calypso/pull/17134 introduced a new bug because of `setRouteMiddleware` called for isomorphic sections. It would dispatch a `ROUTE_SET` action which would reset the page title and meta as defined in the reducer https://github.com/Automattic/wp-calypso/blob/6ec964c85d93bdc7f63400e573a7df6babd64e51/client/state/document-head/reducer.js#L26

Can you elaborate a little on that? Can't we just use our existing isomorphic tools if we're dealing with an isomorphic section?

Do we really need a reducer/Redux state for query args (instead of just getting them from `context`)? And why do we, architecturally, need to store both `initial` and `current` query args?

It'd be nice to see a little more rationale provided on a PR that introduces so many new entities and changes so much code.

**Comment by Tug** [comment]
@ockham Thanks for joining in. I agree the PR description doesn't help understanding what's going on here. Basically it started as a simple fix and evolved to an important refactor of the SSR part as I explained in this comment https://github.com/Automattic/wp-calypso/pull/19036#issuecomment-341728386.

I need to update the PR description to reflect the last changes but basically we don't need to bother with the `document-head` reducer being reset on `ROUTE_SET` anymore as the state is cached and retrieve at the same time as the rendering part now. So even though `setRouteMiddleware` will reset its value before render is called, the entire state will be restored to what it was just when the first rendering ran. Which is more logic than what we have currently imo (trying to reload the state when creating the initial context).

**Comment by Tug** [comment]
> Do we really need a reducer/Redux state for query args (instead of just getting them from context)? And why do we, architecturally, need to store both initial and current query args?

That's just a refactor, I moved those reducers/selectors from one place to another. It seems both (initial and current) are used in the code.

**Comment by seear** [comment]
I did some testing of the /theme and /themes routes. Nice instructions for testing the tags 👍 

Everything seems to be working properly as before. I checked:
```
/themes
/themes/music
/themes/filter/blue
/theme/mood
```
which all cache and use the cache properly.
and 
`/themes?s=twenty`
which should not and does not cache.

The only difference I noticed is that on this branch we are now getting state validation output from the server:
```
  calypso:server-render cache access for key +10s /themes
state validation failed for state: { picks: { featured: { name: 'Featured', description: '' } },
  subject: 
   { announcement: 
      { name: 'Announcement',
        description: 'Themes for weddings, reunions, product launches – whatever you want to announce to family, friends, or customers. Share photos, dates, FAQs, instructions, directions and more.' }...
```

I'm not sure why we don't see that on `master`, because we do the same deserialization.

**Comment by seear** [comment]
> The only difference I noticed is that on this branch we are now getting state validation output from the server...

To clarify, we see the same validation failures at the client, so not related to this PR. Only that we can now see the at the server as well.

**Comment by Tug** [comment]
@seear Indeed I've seen that state validation warning as well but I didn't look at it properly I must admit. Investigating now 👍

**Comment by seear** [comment]
> I've seen that state validation warning as well but I didn't look at it properly I must admit. Investigating now

I don't think it is something that needs too much time spent on it, because it seems like the warning is correct. The weird thing is why we haven't seen it before this PR.

**Comment by Tug** [comment]
👍 I think it's just the `state.themes.themeFilters`'s schema which is not up to date.

**Comment by ockham** [comment]
@Tug Thanks for the explanation, appreciate it!

> Which is more logic than what we have currently imo (trying to reload the state when creating the initial context).

The reason we did this was to re-use themes data fetched from the API for a given theme slug or query IIRC. @seear might be able to confirm :slightly_smiling_face:

**Comment by ockham** [comment]
> In order to fix both issues, the solution adopted in this PR is simple, refactor the caching part of the state so we:
> - always start with an empty state when creating the initial context
> - cache the state only after a page is rendered with SSR (we don't need to cache the state in other cases)
> - retrieve the state and the rendered page at the same time (and overwrite the existing store).

> This approach has a pros:
> - It's cleaner to always start with an empty state and it makes reasoning with SSR simpler
> [...]

Well, to an extent, yes. But caching state on the server has some advantages, too.
I'm starting to understand the underlying issue, but I'm curious -- do you invoke `setRouteMiddleware` more often than once on the server? Is there any need for it being called more than once (per site rendered)?

(A lot of how we conceived SSR to work evolved around the idea that `render` is only called once on the server. By consequence, we should also trigger relevant actions etc only once. If we stick to that constraint and regard a singular SSR'd page (corresponding to a given route and, if necessary, qargs) as our most atomic unit, I think we can continue to cache the relevant data.)

File: server/render/index.js [hunk 9]
```diff
@@ -40,25 +41,64 @@ function bumpStat( group, name ) {
...
+		return pathname;
+	}
+
+	const cachedQueryParams = pick( context.query, context.cacheQueryKeys );
+
+	if ( isEmpty( cachedQueryParams ) ) {
+		return pathname;
+	}
+
+	return pathname + '?' + qs.stringify( cachedQueryParams, { sort: ( a, b ) => a.localCompare( b ) } );
```
Comment (blowery) [inline server/render/index.js]: is `localCompare` supposed to be `localeCompare`?

Comment (ockham) [reply server/render/index.js]: Quite likely :slightly_smiling_face:

Comment (Tug) [reply server/render/index.js]: Good catch! I guess we never actually tested that code since we never had more than 1 cachedQueryParams so far :s

**Comment by Tug** [comment]
> do you invoke setRouteMiddleware more often than once on the server? Is there any need for it being called more than once (per site rendered)?

Well we invoke it during the first rendering then the resulting state is cached. On subsequent renders, the cache will be loaded and setRouteMiddleware will run again, thus the effect will be the same as calling ROUTE_SET twice. A first initial fix was to simply check that the route did not change but there is another problem: it's the `cachedQueryParams` feature I talked about. You can't get the current cache key for the initial state from a request object because you don't know which one of the query parameters are cacheable yet, you need at least the layout for that. 

> Well, to an extent, yes. But caching state on the server has some advantages, too.

I'm not sure what you mean here, it doesn't seem like we cache the state for non-isomorphic routes currently, so the behavior (which page gets cached) hasn't changed as far as I see.

**Comment by ockham** [comment]
>> do you invoke setRouteMiddleware more often than once on the server? Is there any need for it being called more than once (per site rendered)?

> Well we invoke it during the first rendering then the resulting state is cached. On subsequent renders, the cache will be loaded and setRouteMiddleware will run again, thus the effect will be the same as dispatching ROUTE_SET twice.

Hmm, but why is that an issue? We're using the route as our cache key, aren't we? So there shouldn't theoretically be any pollution across cached markups for different routes.

> A first initial fix was to simply check that the route did not change before dispatching ROUTE_SET

I still don't get why we'd have to do that, if we're assigning different cache slots to different routes...

> but there is another problem: the `cachedQueryParams` feature I talked about. You can't get the current cache key for the initial state from a request object because you don't know which one of the query parameters are cacheable yet, you need at least the layout for that.

Why the layout? What we need is some middleware to determine whether the query params are cacheable or not, no?

>> Well, to an extent, yes. But caching state on the server has some advantages, too.

> I'm not sure what you mean here, it doesn't seem like we cache the state for non-isomorphic routes currently, so the behavior (which page gets cached) hasn't changed as far as I see.

Yeah, but we'd be losing it for isomorphic routes, wouldn't we? I'm referring to this line of the PR desc:

> always start with an empty state when creating the initial context

Again, I really appreciate having this discussion -- helps me a lot getting a better understanding for this PR and the underlying issue 👍

**Comment by Tug** [comment]
I do appreciate the discussion as well and I think you are right to try to challenge my approach :)

> Hmm, but why is that an issue? We're using the route as our cache key, aren't we? So there shouldn't theoretically be any pollution across cached markups for different routes.

I'm talking about same routes here.
- Let's load `/log-in` for the first time
- `ROUTE_SET` will be dispatched on the server once (state: `documentHeads: { title: "" }`)
- `renderToString` will be called and the login controller will update the state thanks to [`<DocumentHead ... />`](https://github.com/Automattic/wp-calypso/blob/master/client/login/wp-login/index.jsx#L172) (state: `documentHeads: { title: "Log In" }`
- The state will be cached as `( '/log-in', { documentHeads: { title: "Log In" } } )` basically
- Let's reload `/log-in` now
- The previous state will be retrieved (state: `documentHeads: { title: "Log In" }`)
- `ROUTE_SET` will be dispatched on the server (state: `documentHeads: { title: "" }`)
- `renderToString` won't be called as we will use the cached version of the markup this time
- Result: wrong page title on subsequent calls to `/log-in`.

Now my initial comment should make more sense:
> A first initial fix was to simply check that the route did not change before dispatching ROUTE_SET

> Why the layout? What we need is some middleware to determine whether the query params are cacheable or not, no?

Sorry it's not exactly the layout but the controller actually (the layout middleware `makeLayout` comes just after though). The approach we took is to define `context.cacheQueryKeys` there (in the controller). Currently, the single instance of it we have is in the [login controller](https://github.com/Automattic/wp-calypso/blob/master/client/login/controller.js#L24).
So as you know this part will run [much later](https://github.com/Automattic/wp-calypso/blob/master/server/pages/index.js#L513) than when the context [need to be initialized](https://github.com/Automattic/wp-calypso/blob/master/server/pages/index.js#L154).
So my thinking is that this approach of loading a populated state again before all middlewares is wrong as its equivalent to dispatching all actions twice for this state.

So if we go with my changes we won't lose any state, it will just be used differently.
In the future we will probably want to skip running all those middlewares for cached pages but that's another story.

**Comment by ockham** [comment]
> I'm talking about same routes here.
> - Let's load `/log-in` for the first time
> - `ROUTE_SET` will be dispatched on the server once (state: `documentHeads: { title: "" }`)
> - `renderToString` will be called and the login controller will update the state thanks to [`<DocumentHead ... />`](https://github.com/Automattic/wp-calypso/blob/master/client/login/wp-login/index.jsx#L172) (state: `documentHeads: { title: "Log In" }`
> - The state will be cached as `( '/log-in', { documentHeads: { title: "Log In" } } )` basically
> - Let's reload `/log-in` now
> - The previous state will be retrieved (state: `documentHeads: { title: "Log In" }`)
> - `ROUTE_SET` will be dispatched on the server (state: `documentHeads: { title: "" }`)
> - `renderToString` won't be called as we will use the cached version of the markup this time
> - Result: wrong page title on subsequent calls to `/log-in`.

Okay, thanks for the step-by-step explanation!

I could've sworn I've been bitten by this scenario before (and I thought I'd fixed it at the time, hmm). The fix that seems the most obvious to me is to not only cache the markup, but also the `documentHead` related stuff (title, meta, link), no?

**Comment by ockham** [comment]
Hmm, I might've misread that a bit upon first read. I'll give it some more thought.

Still inclined to think that avoiding to fire `ROUTE_SET` seems like the right strategy -- that, or more generally making sure that we end up with the right state either way.

**Comment by ockham** [comment]
> So my thinking is that this approach of loading a populated state again before all middlewares is wrong as its equivalent to dispatching all actions twice for this state.

I'm not sure I agree. This was sort of a driving design principle for SSR caching -- see my example of the theme showcase which would either fire an API request to fetch themes data each time, or just re-use cached state. 

Granted, in order for this to work, we need to make sure all those actions _aren't_ fired when we have cached state available.

**Comment by ockham** [comment]
BTW whatever we end up doing, in order to avoid future confusion around fixes breaking stuff, we should really cover the cases we're trying to fix here with tests...

**Comment by ockham** [comment]
> Sorry it's not exactly the layout but the controller actually (the layout middleware `makeLayout` comes just after though). The approach we took is to define `context.cacheQueryKeys` there (in the controller). Currently, the single instance of it we have is in the [login controller](https://github.com/Automattic/wp-calypso/blob/master/client/login/controller.js#L24).

> So as you know this part will run [much later](https://github.com/Automattic/wp-calypso/blob/master/server/pages/index.js#L513) than when the context [need to be initialized](https://github.com/Automattic/wp-calypso/blob/master/server/pages/index.js#L154).

Okay, so I'd say the problem we're facing is the order of invocation of setting the `initialState` from cache, and defining which query keys to take into account for cache keys. 

By definition, we're restoring `initialState` from cache early on, before we run through the section's individual middlewares. So maybe we should hard-code a list of query keys in `sections.js` after all, since that is guaranteed to be evaluated earlier.

File: client/state/ui/guided-tours/selectors/index.js [hunk 10]
```diff
@@ -22,7 +22,8 @@ import debugFactory from 'debug';
  * Internal dependencies
  */
 import { GUIDED_TOUR_UPDATE, ROUTE_SET } from 'state/action-types';
-import { getInitialQueryArguments, getSectionName } from 'state/ui/selectors';
+import { getSectionName } from 'state/ui/selectors';
+import getInitialQueryArguments from 'state/selectors/get-initial-query-arguments';
```
Comment (ockham) [inline client/state/ui/guided-tours/selectors/index.js]: Should be `getInitialQueryArguments from '{ state/selectors }`.

But I really kinda distrust these selectors altogether. What does initial vs current qargs mean? That just adds confusion to something that's confusing enough IMO.

Comment (ockham) [reply client/state/ui/guided-tours/selectors/index.js]: Asked for motivation: https://github.com/Automattic/wp-calypso/pull/6417#issuecomment-343751609

**Comment by ockham** [comment]
BTW why are we calling `setRoute` [on the server](https://github.com/Automattic/wp-calypso/pull/17134/files#diff-336276ec8a6d8fa6e6ecd99de5a7bab3R39) at all?

**Comment by Tug** [comment]
> see my example of the theme showcase which would either fire an API request to fetch themes data each time, or just re-use cached state.

**Comment by Tug** [comment]
> see my example of the theme showcase which would either fire an API request to fetch themes data each time, or just re-use cached state.

I understand why you needed that now. So why not just skip all middlewares when we hit the cache and render the page directly?

**Comment by Tug** [comment]
> BTW why are we calling setRoute on the server at all?

It's used for [customized layout for dops clients](https://github.com/Automattic/wp-calypso/blob/master/client/layout/logged-out.jsx#L84) where we need the query arguments for SSR.

It will be used for other purposes imo. Having access to the current route before page.js boots can be useful in some cases, like redirecting isomorphically in some controllers ( ex: `if ( getCurrentUser( state ) && getCurrentRoute( '/me' ) ) return context.redirect( '/log-in' );` )

**Comment by Tug** [comment]
> By definition, we're restoring initialState from cache early on, before we run through the section's individual middlewares. So maybe we should hard-code a list of query keys in sections.js after all, since that is guaranteed to be evaluated earlier.

I agree here, having those defined in the controllers doesn't make much sense since they act on the cache there should be a reliable way to determine if a query param is cacheable or not

**Comment by Tug** [comment]
@ockham would that change work for you https://github.com/Automattic/wp-calypso/pull/19036/commits/7916a15857eec0ddb18bce5e81c0ade690d29826#diff-336276ec8a6d8fa6e6ecd99de5a7bab3R30 ?

It renders a page early so we don't go through all the middlewares again when a page is cached.

> BTW whatever we end up doing, in order to avoid future confusion around fixes breaking stuff, we should really cover the cases we're trying to fix here with tests...

Good point! Are you thinking of unit tests inside calypso? I would need to think a bit about how to achieve that. Maybe if we call `serverRender` twice on a few isomorphic routes to assert that the `context.renderedLayout` hasn't changed. We could probably complete that with e2e tests.

**Comment by yurynix** [comment]
I've squashed this branch for a more easier rebase and rebased it.
I've fixed a few style issues mentioned by @ockham ✔️ 
I've tested and the code worked as described at the description ✔️ 

I do agree we might want to have more tests for that PR, however, this PR
is already quite big and introduces an improvement over what we currently have,
so IMO we should merge that and follow up with an additional PR with tests ( in whatever strategy unit / e2e ).

So, IMO, LGTM 🚢

**Comment by ockham** [comment]
> @ockham would that change work for you 7916a15#diff-336276ec8a6d8fa6e6ecd99de5a7bab3R30 ?

> It renders a page early so we don't go through all the middlewares again when a page is cached.

@Tug Yeah, that's better I think. Thanks!

@yurynix I'd be a bit more comfortable if we only merged the changes that are actually need to fix the issue here -- IIUC, most of the 22 files touched are more of a refactor that isn't directly related to the issue?

**Comment by Tug** [comment]
@ockham https://github.com/Automattic/wp-calypso/pull/20255 was merged first with only the refactors on the state, I rebased and merged this one with just the SSR changes, let's keep an eye on it for any weird behaviour ;)
